File: crates/prova-core/tests/requires.rs
use std::path::PathBuf;

use prova_core::{run_path, NullReporter};

/// `requires` gates on capability availability: an unavailable capability SKIPS the unit (not
/// fails), and that skip cascades to dependents. An available capability (or no requirement) runs.
#[test]
fn requires_skips_on_missing_capability() {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("testdata/requires.lua");
    let mut reporter = NullReporter;
    let summary = run_path(&path, &mut reporter).expect("run requires.lua");

    // "needs sh" + "no requirements" + "needs kafka (compiled in)" pass; the two missing-tool gates
    // + the dependent skip. (The kafka assertion assumes the default-featured test build.)
    assert_eq!(summary.passed, 3, "passed");
    assert_eq!(
        summary.failed, 0,
        "failed (nothing ever fails on a missing capability)"
    );
    assert_eq!(summary.skipped, 3, "skipped");
}

File: crates/prova-core/testdata/requires.lua
-- Unavailable capability → the unit is SKIPPED, not failed (its body never runs).
prova.test("needs a missing tool", { requires = { "definitely_not_a_real_tool_xyzzy" } }, function(t)
  error("must never run")
end)

-- Available capability (sh is on PATH) → runs normally.
prova.test("needs sh which is present", { requires = { "sh" } }, function(t)
  t:expect(1):equals(1)
end)

prova.test("no requirements runs", function(t)
  t:expect(true):is_true()
end)

-- A native-client capability is available when its feature is compiled into the build (sqlite is the
-- embedded database, still bundled), even though there is no `sqlite` binary on PATH — the unified
-- gate checks the build, not PATH. (Before unification this spuriously skipped, looking for a binary.)
prova.test("needs the sqlite client (compiled in)", { requires = { "sqlite" } }, function(t)
  t:expect(1):equals(1)
end)

-- A skipped-by-requires unit cascades to its dependents (skip, not fail).
local gated = prova.test("gated on a missing tool", { requires = { "definitely_not_a_real_tool_xyzzy" } }, function(t)
  error("must never run")
end)
prova.test("depends on the gated unit", { depends_on = { gated } }, function(t)
  error("must never run — upstream was skipped")
end)
